README.md
<img src="rdnb_logo.png" alt="rdnb" />

An R interface to the Deutsche Nationalbibliothek (German National Library) API

[![Build Status](https://api.travis-ci.org/chgrl/rdnb.png)](https://travis-ci.org/chgrl/rdnb)
![downloads](http://cranlogs.r-pkg.org/badges/grand-total/rdnb)


### About the package

`rdnb` is an R wrapper for the Deutsche Nationalbibliothek (German National Library) API, available at http://www.dnb.de. As the German central archival library, the German National Library is collecting, archiving, bibliographically classifying all German and German-language publications, foreign publications about Germany, translations of German works, and the works of German-speaking emigrants published abroad between 1933 and 1945.

All bibliographic data of the German National Library are provided free of charge and can be freely re-used under "Creative Commons Zero" ([CC0 1.0](https://creativecommons.org/publicdomain/zero/1.0/deed.de)) terms. The metadata and online interfaces are provided with no guarantee of their being continuous, punctual, error-free or complete, or of their not infringing the rights of third parties (e.g. personal rights and copyright).


#### Source and more details:

[About the DNB](https://www.dnb.de/EN/Ueber-uns/ueberUns_node.html); 
[About the interface](https://www.dnb.de/EN/Professionell/Metadatendienste/Datenbezug/SRU/sru_node.html)


### Package install

`rdnb` is available on CRAN and GitHub. Install from CRAN:


```r
install.packages("rdnb")
```

To install the development version from GitHub the `devtools`-package is required.


```r
devtools::install_github("chgrl/rdnb")
```


### Load the package


```r
library(rdnb)
```


### Search the DNB catalogue

`rdnb` provides two approaches to expose a search. For common search requests use the simple `dnb_search` function. More complex requests are possible using the `dnb_advanced` function.

#### Simple search - `dnb_search`

##### Search for title

Use the `title` parameter to search in titles, including subtitle, short title, volume title, etc.


```r
single.title <- dnb_search(title="katze")
```

A vector of strings combines search terms. Strings in a vector are "OR"-combined by default. Logical operators can be changed by starting a string with `+` or `-`, thus the following code searches for titles containing one of "katze" or "kater", "maus" but not "hund".


```r
multi.title <- dnb_search(title=c("katze", "kater", "+maus", "-hund"))
```

##### Search for author

The `author` parameter offers a search for one or a set of authors. Logical operators can be set as explained above.


```r
single.author.1 <- dnb_search(author="kern")
single.author.2 <- dnb_search(author="kern, ingolf")
single.author.3 <- dnb_search(author="ingolf kern")
author.or.author <- dnb_search(author=c("kern", "locke"))
author.and.author <- dnb_search(author=c("kern", "+locke"))
author.not.author <- dnb_search(author=c("kern", "-locke"))
```

##### Search for publication year

Use the `year` parameter to limit the year(s) of publication. Set `year` to a single integer value to specify a single year. Set the parameter to a sequence to specify start and end. Or set the parameter to a vector of integer values to specify a set of years.


```r
single.year <- dnb_search(title="katze", year=2015)
sequence.of.years <- dnb_search(title="katze", year=2010:2015)
set.of.years <- dnb_search(title="katze", year=c(2010:2013, 2015))
```

##### Search for publisher

Publisher name and/or location may be limited by the `publisher` parameter. Logical operators can be set as explained above.


```r
single.publisher <- dnb_search(title="katze", publisher="kiepenheuer")
single.publisher.location <- dnb_search(title="katze", publisher="*verlag leipzig*")
pub.or.pub <- dnb_search(title="katze", publisher=c("kiepenheuer", "piper"))
pub.and.pub <- dnb_search(title="katze", publisher=c("*kinder*", "+*berlin*"))
pub.not.pub <- dnb_search(title="katze", publisher=c("pi*", "-piper"))
```

##### Search for keyword

Keyword (subjects, persons, locations, organisations, etc.) search is available using the `keyword` parameter. Logical operators can be set as explained for title search.


```r
single.keyword <- dnb_search(keyword ="katze")
keyword.or.keyword <- dnb_search(keyword=c("katze", "hund"))
keyword.and.keyword <- dnb_search(keyword=c("katze", "+hund"))
keyword.not.keyword <- dnb_search(keyword=c("katze", "-hund"))
```

##### Search for type of publication

The DNB discriminates between several publication types:

  * articles (Artikel)
  * manuscript (Manuskripte)
  * biographicaldoc (Lebensdokumente)
  * letters (Briefe)
  * bequest (Nachlässe)
  * collections (Sammlungen)
  * books (Bücher)
  * brailles (Blindenschriften)
  * maps (Karten)
  * discs (Elektronische Datenträger)
  * dissertations (Hochschulschriften)
  * online (Online Ressourcen)
  * films (Filme/Hörbücher)
  * microfiches (Mikrofilm)
  * multimedia (Medienkombinationen)
  * music (Musiktonträger)
  * scores (Noten)
  * serials (Zeitschriften)
  * persons (Personen)
  * subjects (Sachbegriffe)
  * corperations (Organisationen)
  * works (Werke)
  * events (Ereignisse)
  * geographics (Geografika)
  
The `type` parameter can be specified as single string value (named above) or a vector of string values, that will be combined by "OR". Short terms of the available types usually work.


```r
single.type <- dnb_search(title="katze", type="books")
single.type.part <- dnb_search(title="katze", type="bio") # biographicaldoc
set.of.types <- dnb_search(title="katze", type=c("books", "articles", "online"))
```

##### Search for language

The DNB also archives publications in other languages than German, e.g. when they are about Germany or publicated in Germany. The language(s) can be filtered by ISO 639-2/B [code](https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes), like "ger" (German), "eng" (English), "fre" (French), "spa" (Spanish) or "gsw" (Swiss German). A vector of string values is combined by "OR".


```r
single.language <- dnb_search(title="cat", language="eng")
set.of.languages <- dnb_search(title=c("cat", "katze"), language=c("eng", "ger"))
```

##### Change limit of results

Use the `limit` parameter to set the number and optionally the starting point of the results returned. A single integer value (possible values: 1-100) specifies the number of results only. Use a vector of two integer values to specify the number of results (first value of vector) and the first result (1 or higher). Set `limit` to "all" for a complete list of results.


```r
first.result <- dnb_search(title="katze", limit=1)
five.results.starting.with.the.21st <- dnb_search(title="katze", limit=c(5, 21))
all.results <- dnb_search(title="katze", year=2016, limit="all")
```

##### Results

Per default no results are printed. Set `print` to `TRUE` to print the `data.frame` with the results.


```r
cats <- dnb_search(title="katze", limit=3, print=TRUE)
#>           id                        link
#> 1 1116362198 http://d-nb.info/1116362198
#> 2 1120516579 http://d-nb.info/1120516579
#> 3 1125403497 http://d-nb.info/1125403497
#>                                               author
#> 1 Teckentrup, Britta (ill); Teckentrup, Britta (aut)
#> 2                       Frey, Alexander Moritz (aut)
#> 3                            Krokowski, Carola (aut)
#>                              title             subtitle
#> 1 Alles Natur - Meine kleine Katze                 <NA>
#> 2            Birl, die kühne Katze          Ein Märchen
#> 3   Blutwerte - Pferd, Hund, Katze Blutwerte verstehen!
#>                        publisher year language          isbn
#> 1           Ars Edition, München 2017      ger 9783845815886
#> 2       Elsinor Verlag, Coesfeld 2017      ger 9783942788373
#> 3 Igelsburg Verlag, Habichtswald 2017      ger 9783941933200
#>                                                              price pages
#> 1   Pappe : EUR 7.99 (DE), EUR 8.30 (AT), CHF 11.90 (freier Preis)    16
#> 2            Broschur : circa EUR 12.80 (DE), circa EUR 13.20 (AT)    88
#> 3 : EUR 29.95 (DE), EUR 29.95 (DE), EUR 30.80 (AT), EUR 30.30 (AT)     2
#>              format                  edition keyword  toc
#> 1 17.3 cm x 17.3 cm                     <NA>    <NA> <NA>
#> 2     19 cm x 12 cm 1. Auflage, neue Ausgabe    <NA> <NA>
#> 3              <NA>               1. Auflage    <NA> <NA>
#>                                                                                               description
#> 1 http://deposit.d-nb.de/cgi-bin/dokserv?id=992efedfe14a4b06ad7dc4f344cd9e87&prov=M&dok_var=1&dok_ext=htm
#> 2                                                                                                    <NA>
#> 3                                                                                                    <NA>
#>                                                             cover
#> 1 https://portal.dnb.de/opac/mvb/cover.htm?isbn=978-3-8458-1588-6
#> 2 https://portal.dnb.de/opac/mvb/cover.htm?isbn=978-3-942788-37-3
#> 3 https://portal.dnb.de/opac/mvb/cover.htm?isbn=978-3-941933-20-0
```

The following data is stored in the `data.frame` (if available):

| Variable | Description |
| --- | --- |
| id | DNB record ID |
| link | link to record |
| author | author(s) |
| title | main title |
| subtitle | subtitle |
| publisher | publisher |
| year | year of publication |
| language | language (abbreviation) |
| isbn | ISBN (13-digit or 10-digit) |
| price | price (as string, usually prices for DE, AT and CH) |
| pages | pages |
| format | size and weight (partial) |
| edition | edition |
| keyword | keyword(s) |
| toc | link to table of contents |
| description | link to description |
| cover | link to cover image |

#### Advanced search - `dnb_advanced`

`dnb_advanced` allows for complex requests using the Contextual Query Language (CQL). See the [DNB advanced search help pages](https://www.dnb.de/expertensuche) for available indices and a list of examples.

The following advanced search returns a list of german or english children's books titled with 'cat', excluding titles containing 'dog', since the year 2001:


```r
cats <- dnb_advanced("( (tit=katze OR tit=kater NOT tit=hund) OR (tit=cat NOT tit=dog) ) AND jhr>2000 AND mat=books AND (spr=ger OR spr=eng) AND sgt=K", limit="all")
```


### Utils

#### Number of records

Use `n_rec` to get the total number of records found by a request:


```r
n_rec(cats)
#> [1] 5895
```

#### Print query

To print the query used for a DNB-search, call `print_query`:


```r
print_query(cats)
#> [1] "(tit=katze)"
```

You can use this query (or an edited version) directly with the `dnb_advanced` function:


```r
cat.q <- print_query(cats)
cat.q <- gsub("jhr>2000", "jhr>=2015", cat.q) # change year
dnb_advanced(cat.q)
```


inst/NEWS
rdnb 0.1-6 [2023-02-17]
-----------------------
* bug fixed: potential error in case of empty sublists in dnb-data

rdnb 0.1-5 [2022-05-12]
-----------------------
* bug fixed: dnb_advanced with limit="all" crashed

rdnb 0.1-4 [2021-04-17]
-----------------------
* no access token required anymore!
* links fixed
* missing dependency added

rdnb 0.1-3 [2019-04-05]
-----------------------
* parsing error fixed (changes in XML structure)
* improvements in clearing out and cleaning of data

rdnb 0.1-2 [2018-09-15]
-----------------------
* toggle between hide and show url via option rdnb_show_url
  (for debugging)
* new function save_token to save token to file
* vignette update
* DESCRIPTION fixed for CRAN

rdnb 0.1-1 [2017-02-14]
-----------------------
* data cleaning option added
* progress bar added (for limit="all" only)
* limitation for number of results (1-100) removed
* NOT-operator support added to publisher search
* new function "n_rec" added - get number of records
* new function "print_query" added - print search query
* other small fixes and improvements

rdnb 0.1-0 [2016-03-30]
-----------------------
* This is the first version


R/zzz.R
.onAttach <- 
function(libname, pkgname) {
    ver <- read.dcf(file=system.file("DESCRIPTION", package=pkgname), fields="Version")
    packageStartupMessage(" ")
    packageStartupMessage(paste("This is", pkgname, ver))
    packageStartupMessage(" ")
    packageStartupMessage("Type changes(\"rdnb\") to see changes/bug fixes, help(rdnb) for documentation")
    packageStartupMessage("or citation(\"rdnb\") for how to cite rdnb.")
    packageStartupMessage(" ")
}

.onLoad <- 
function(libname, pkgname) {
  options(rdnb_show_url=FALSE)
}

#' @title View changes notes.
#' @description \code{changes} brings up the NEWS file of the package. 
#'
#' @param pkg Set to the default "rdnb". Other packages make no sense.
#' @export
#' @examples
#' \dontrun{
#' changes()
#' }
changes <- 
function(pkg="rdnb") {
    if(pkg=="rdnb") file.show(file.path(system.file(package="rdnb"), "NEWS"))
}


R/rdnb.R
#' @title R Interface to the Deutsche Nationalbibliothek (German National Library) API
#' @description A wrapper for the Deutsche Nationalbibliothek (German National Library) API, available at \url{https://www.dnb.de/EN/Home/home_node.html}. The German National Library is the German central archival library, collecting, archiving, bibliographically classifying all German and German-language publications, foreign publications about Germany, translations of German works, and the works of German-speaking emigrants published abroad between 1933 and 1945.
#' @name rdnb
#' @docType package
#' @details All bibliographic data of the German National Library are provided free of charge and can be freely re-used under "Creative Commons Zero" (CC0 1.0) terms. The metadata and online interfaces are provided with no guarantee of their being continuous, punctual, error-free or complete, or of their not infringing the rights of third parties (e.g. personal rights and copyright).
#' @references About the DNB: \url{https://www.dnb.de/EN/Ueber-uns/ueberUns_node.html}; about the interface: \url{https://www.dnb.de/EN/Professionell/Metadatendienste/Datenbezug/SRU/sru_node.html}
#' @importFrom httr GET
#' @importFrom httr content
#' @importFrom stringr str_split
#' @importFrom xml2 as_list
#' @importFrom xml2 read_xml
#' @importFrom stats setNames
#' @import brew
#' @import grDevices
#' @import methods
#' @import utils
#' @aliases rdnb rdnb-package
NULL


R/gets.R
#' @title Search the DNB catalogue - simple search
#' @description \code{dnb_search} exposes a search in the DNB catalogue. 
#' @param title the title (including subtitle, short title, volume title, etc.); optional single string value or vector of strings.
#' @param author the author(s); optional single string value or vector of strings.
#' @param year the year of publishing; optional single integer value or vector of integers.
#' @param publisher the publisher (publisher name and/or location); optional single string value or vector of strings.
#' @param keyword one or a set of keywords describing the work (subjects, persons, locations, organisations, etc.); optional single string value or vector of strings.
#' @param type the type of publication (optional), one or a vector of \code{articles}, \code{manuscript}, \code{biographicaldoc}, \code{letters}, \code{bequest}, \code{collections}, \code{books}, \code{brailles}, \code{maps}, \code{discs}, \code{dissertations}, \code{online}, \code{films}, \code{microfiches}, \code{multimedia}, \code{music}, \code{scores}, \code{serials}, \code{persons}, \code{subjects}, \code{corperations}, \code{works}, \code{events}, \code{geographics}.
#' @param language the language of the work by ISO 639-2/B code (\url{https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes}); single string value or vector of strings.
#' @param limit number and (optional) starting point of results returned; single integer value (number of results), vector of two integer values (number of results and first result, >=1) or \code{"all"} for a complete list of results.
#' @param clean if \code{TRUE} (the default), the results are cleaned (see \code{\link{dnb_advanced}} for details).
#' @param print if \code{TRUE} the search results are printed (default is \code{FALSE}).
#' @return A list of results with metadata.
#' @details to do
#' @export
#' @examples
#' \dontrun{
#' # title search
#' single.title <- dnb_search(title="katze")
#' multi.title <- dnb_search(title=c("katze", "kater", "+maus", "-hund"))
#'
#' # author search
#' single.author <- dnb_search(author="kern")
#' author.or.author <- dnb_search(author=c("kern", "locke"))
#' author.and.author <- dnb_search(author=c("kern", "+locke"))
#' author.not.author <- dnb_search(author=c("kern", "-locke"))
#'
#' # publication year 
#' single.year <- dnb_search(title="katze", year=2015)
#' sequence.of.years <- dnb_search(title="katze", year=2010:2015)
#' set.of.years <- dnb_search(title="katze", year=c(2010:2013, 2015))
#'
#' # publisher search
#' single.publisher <- dnb_search(title="katze", publisher="kiepenheuer")
#' set.of.publishers <- dnb_search(title="katze", publisher=c("kiepenheuer", "piper"))
#'
#' # keyword search
#' single.keyword <- dnb_search(author="kern")
#' keyword.or.keyword <- dnb_search(keyword=c("katze", "hund"))
#' keyword.and.keyword <- dnb_search(keyword=c("katze", "+hund"))
#' keyword.not.keyword <- dnb_search(keyword=c("katze", "-hund"))
#'
#' # type search
#' single.type <- dnb_search(title="katze", type="books")
#' set.of.types <- dnb_search(title="katze", type=c("books", "articles", "online"))
#'
#' # language search
#' single.language <- dnb_search(title="cat", language="eng")
#' set.of.languages <- dnb_search(title=c("cat", "katze"), language=c("eng", "ger"))
#'
#' # change limit of results
#' first.result <- dnb_search(title="katze", limit=1)
#' 5.results.starting.with.the.21st <- dnb_search(title="katze", limit=c(5, 21))
#' all.results <- dnb_search(title="katze", limit="all")
#' }
dnb_search <- function(title, author, year, publisher, keyword, type, language, limit=10, clean=TRUE, print=FALSE) {		
	# init query
	query <- ""
	
	# prepare title
	if(!missing(title)) {
		title <- paste0("tit=", title, collapse=" OR ")
		title <- gsub("OR tit=+", "AND tit=", title, fixed=TRUE)
		title <- gsub("OR tit=-", "NOT tit=", title, fixed=TRUE)
		if(substr(title, 1, 5)=="tit=+" || substr(title, 1, 5)=="tit=-") stop("Do not use '+' or '-' in case of a single search string or with first string of a vector")
		query <- paste0("(", title, ")")
	}
	
	# prepare author
	if(!missing(author)) {
		author <- paste0("atr=", author, collapse=" OR ")
		author <- gsub("OR atr=+", "AND atr=", author, fixed=TRUE)
		author <- gsub("OR atr=-", "NOT atr=", author, fixed=TRUE)
		if(substr(author, 1, 5)=="atr=+" || substr(author, 1, 5)=="atr=-") stop("Do not use '+' or '-' in case of a single search string or with first string of a vector")
		if(query=="") query <- paste0("(", author, ")")
		else query <- paste(query, paste0("(", author, ")"), sep=" AND ") 
	}
	
	# prepare year
	if(!missing(year)) {
		if(length(year)>1 && (tail(year, 1)-year[1]+1)==length(year)) {
			year <- paste0("jhr>=", year[1], " AND jhr<=", tail(year, 1))
		} else {
			year <- paste0("jhr=", year, collapse=" OR ")
		}
		if(query=="") query <- paste0("(", year, ")")
		else query <- paste(query, paste0("(", year, ")"), sep=" AND ")
	}
	
	# prepare publisher
	if(!missing(publisher)) {
		publisher <- paste0("vlg=", publisher, collapse=" OR ")
		publisher <- gsub("OR vlg=+", "AND vlg=", publisher, fixed=TRUE)
		publisher <- gsub("OR vlg=-", "NOT vlg=", publisher, fixed=TRUE)
		if(substr(publisher, 1, 5)=="vlg=+" || substr(publisher, 1, 5)=="vlg=-") stop("Do not use '+' or '-' in case of a single search string or with first string of a vector")
		if(query=="") query <- paste0("(", publisher, ")")
		else query <- paste(query, paste0("(", publisher, ")"), sep=" AND ") 
	}
	
	# prepare keyword
	if(!missing(keyword)) {
		keyword <- paste0("sw=", keyword, collapse=" OR ")
		keyword <- gsub("OR sw=+", "AND sw=", keyword, fixed=TRUE)
		keyword <- gsub("OR sw=-", "NOT sw=", keyword, fixed=TRUE)
		if(substr(keyword, 1, 5)=="sw=+" || substr(keyword, 1, 5)=="sw=-") stop("Do not use '+' or '-' in case of a single search string or with first string of a vector")
		if(query=="") query <- paste0("(", keyword, ")")
		else query <- paste(query, paste0("(", keyword, ")"), sep=" AND ")
	}
	
	# prepare type
	if(!missing(type)) {
		avail.types <- c("articles", "manuscript", "biographicaldoc", "letters", "bequest", "collections", "books", "brailles", "maps", "discs", "dissertations", "online", "films", "microfiches", "multimedia", "music", "scores", "serials", "persons", "subjects", "corperations", "works", "events", "geographics")
		type <- avail.types[pmatch(type, avail.types)]		
		type <- paste0("mat=", type, collapse=" OR ")
		if(query=="") query <- paste0("(", type, ")")
		else query <- paste(query, paste0("(", type, ")"), sep=" AND ")
	}
	
	# prepare language
	if(!missing(language)) {
		language <- paste0("spr=", language, collapse=" OR ")
		if(query=="") query <- paste0("(", language, ")")
		else query <- paste(query, paste0("(", language, ")"), sep=" AND ")
	}
	
	# call dnb_advanced
	df <- dnb_advanced(query=query, limit=limit, clean=clean, print=FALSE)
  
  # return
  if(print) print(df)
  invisible(df)
}


#' @title Search the DNB catalogue - advanced search
#' @description \code{dnb_search} exposes a search in the DNB catalogue, expressed in the DNB query language. 
#' @param query the search query, expressed in the DNB query language; single string value. 
#' @param limit number and (optional) starting point of results returned; single integer value (number of results), vector of two integer values (number of results and first result, >=1) or \code{"all"} for a complete list of results.
#' @param clean if \code{TRUE} (the default), the results are cleaned (see below for details).
#' @param print if \code{TRUE} the search results are printed (default is \code{FALSE}).
#' @return A \code{data.frame} of results with metadata.
#' @details \emph{Cleaning of results}
#'
#' To harmonise the results, some clutter is deleted and abbreviations frequently used in the dnb catalogue are replaced by full words. Here is a list of all replacements:
#' \tabular{lll}{
#' Variable \tab Searched \tab Replaced by \cr
#' complete dataset \tab \\u0098 \tab [deleted] \cr
#' complete dataset \tab \\u009c \tab [deleted] \cr
#' complete dataset \tab ,, \tab , \cr
#' complete dataset \tab .. \tab . \cr
#' complete dataset \tab ;; \tab ; \cr
#' year \tab [string] \tab [numeric] \cr
#' pages \tab  S. \tab [deleted] \cr
#' pages \tab  Seiten \tab [deleted] \cr
#' pages \tab [ \tab [deleted] \cr
#' pages \tab ] \tab [deleted] \cr
#' publisher \tab Verl. \tab Verlag \cr
#' publisher \tab verl. \tab verlag \cr
#' publisher \tab [ \tab [deleted] \cr
#' publisher \tab ] \tab [deleted] \cr
#' edition \tab Aufl. \tab Auflage \cr
#' edition \tab aufl. \tab auflage \cr
#' edition \tab Orig. \tab Original \cr
#' edition \tab Ed. \tab Edition \cr
#' edition \tab ed. \tab edition \cr
#' edition \tab Ausg. \tab Ausgabe \cr
#' edition \tab ausg. \tab ausgabe \cr
#' edition \tab Nachdr. \tab Nachdruck \cr
#' edition \tab Bibliogr. \tab Bibliografie \cr
#' edition \tab [ \tab [deleted] \cr
#' edition \tab ] \tab [deleted] \cr
#' edition \tab [x]., \tab [x]. \cr 
#' price \tab kart. \tab Kartoniert \cr
#' price \tab Gb. \tab Gebunden \cr
#' price \tab Spiralb. \tab Spiralbindung \cr
#' price \tab Pb. \tab Paperback
#' }
#' @export
#' @examples
#' \dontrun{
#' # german books titled with 'cat' (male or female), 
#' # excluding titles containing dogs, since the year 2001
#' cats <- dnb_advanced("(tit=katze OR tit=kater NOT tit=hund) AND jhr>2000 AND mat=books AND spr=ger")
#' }
dnb_advanced <- function(query, limit=10, clean=TRUE, print=FALSE) {		
	# prepare limit
	if(any(limit=="all")) {
		lim <- 100
		strt <- 1
	} else if(is.numeric(limit)) {
		if(length(limit)==1) {
			lim <- limit
			strt <- 1
		} else if(length(limit)==2) {
			lim <- limit[1]
			strt <- limit[2]
		} else stop("cannot read 'limit'")
	} else stop("cannot read 'limit'")
	
	# make request
  req <- dnb_get_url(path="sru/dnb", query=query, limit=lim, start=strt)
  raw <- dnb_parse(req)$searchRetrieveResponse
  
  # print number of records
	nrec <- as.numeric(raw[["numberOfRecords"]])
	if(nrec==0) {
		message("no records found")
		return(NULL)
	}
  
  # convert
  df <- dnb_to_df(raw, clean=clean)
  
  # loop request for more than 100 records
	nend <- NULL
	if(any(limit=="all") || lim>100) {
		if(any(limit=="all")) {
		  nend <- nrec
		} else {
			nend <- lim
			lim <- 100
		}
		strt <- as.numeric(raw[["nextRecordPosition"]])
		pb <- txtProgressBar(min=0, max=nend, style=3)
		repeat{
		  if(strt>nend) break
			req <- dnb_get_url(path="sru/dnb", query=query, limit=lim, start=strt)
			raw <- dnb_parse(req)$searchRetrieveResponse
			df_add <- dnb_to_df(raw, clean=clean)
			df <- rbind(df, df_add)
			strt <- as.numeric(raw[["nextRecordPosition"]])
			if(length(strt) == 0) break
			setTxtProgressBar(pb, strt)
		}
		setTxtProgressBar(pb, nend)
		close(pb)
		if(!any(limit=="all")) df <- df[1:nend,]
	}
	
	# print number of records
	if(any(limit=="all")) message(nrec, " records found")
	else {
		if(!is.null(nend)) message(nrec, " records found (request limited to ", nend, " records)")
		else message(nrec, " records found (request limited to ", lim, " records)")
	}
	# add metadata
	attr(df, "number_of_records") <- nrec
	attr(df, "query") <- unlist(raw[["echoedSearchRetrieveRequest"]][["query"]])
  
  # return
  if(print) print(df)
  invisible(df)
}
